### ISSUE TITLE
Support mouselook when no VR headset available

### ISSUE THREAD

**Issue by bcampbell** [bjornbytes/lovr#23]
I've started on some fake headset/controller support. It uses mouselook and WASD to move about.
There's a fake controller, locked to the head position/orientation, with left mouse button acting as trigger.

I'm working in this branch: https://github.com/bcampbell/lovr/tree/fake-headset

Not quite ready for a pull request, but I'm pleasantly surprised how little I've had to change.
The changes:
- new set of `lovrHeadset` functions in `headset/fake.c`, analogous to the webvr/openvr split.
- currently the headset implementation is hard-coded at compile-time (like openvr/webvr)
- I added a new lovrHeadsetUpdate() function, which is called from the main loop in `boot.lua`.
- The fake headset needs the GLFW window pointer. I've fudged in a access call in `src/graphics`, and a special case in `boot.lua` to only initialise the headset _after_ window creation. (ugh)

TODO:
- pick appropriate headset implementation at runtime. Some sort of indirection - either within `src/headset`, or in `api/headset.c`.
- think about better methods for headset to get the window ptr (maybe some kind of hook where the headset module can be informed upon window creation/destruction...). My aim is to put the module initialisation in `boot.lua` back how it was.
- tighten up handling of velocity, sit/stand transform...
- the fake controller needs a lot more work - what's there now is a nasty 2minute hack.
- tidy it all up and submit a pull request

**Comment by bjornbytes** [comment]
Awesome!  A while back I had a [struct of function pointers](https://github.com/bjornbytes/lovr/blob/666a3a23d78aad95900d22400b721d191471e9fe/src/headset/headset.h#L34-L59) for the headset interface.  Ended up switching to swapping out implementation files because it reduced a lot of duplication, but it could make sense to go back to the function pointer approach to allow dynamically switching between headset implementations.

**Comment by bjornbytes** [comment]
I've also run into the issue about needing to get the window to the headset before, and haven't come up with a good solution yet.  The current approach is to just delay anything that needs the graphics module for later (the OpenVR implementation waits for renderTo to initialize its texture).  A similar hack that could be used for the fake headset is in the `lovrHeadsetUpdate` function.  It could see if `lovrGraphicsGetWindow` exists and bind the GLFW callbacks if they haven't been bound yet.  That's is still really awkward though...

I think it's kinda cool how it's possible to use `lovr.headset` without `lovr.graphics` right now (except for renderTo) so you can still get tracking info without a window created, but it would be okay if that's no longer the case.

**Comment by bcampbell** [comment]
I'll have a bit more of a play about with the window-binding issue and see if I can think of anything more elegant. I need to sort something out anyway, as it currently coredumps upon exit :- )
Will keep you posted here.

**Comment by shakesoda** [comment]
the picking headset at runtime issue is relevant for #19 also

BTW, this is useful for development even when you *do* have a headset. spare some dev headaches

**Comment by bcampbell** [comment]
I figured that might be the case. I really don't know a lot about the current landscape of VR SDKs - what ones are out there and what platforms (headset/OSes) do they support? Is openvr an attempt to come up with a unifying front end to them all?
(myself, I'm particularly interested in the lower end, untethered headsets, even if they are so often tied up with the general ickiness of coding on android ;- )

**Comment by shakesoda** [comment]
the ones I'm aware of:

1) openvr - native for vive and some other headsets (pimax for example), also works with rift. intended to be more-or-less universal but this comes with some caveats (burns a core and doesn't shutdown properly with rift, for example). works on windows/mac/linux
2) libovr - native for rift, supports everything the hw can do. this namely means better haptics and some oculus touch specific stuff I don't think you can get with openvr. windows-only
3) VrApi - gear vr/oculus go native. android-only
4) webvr - native for nothing, but should work universally. comes with the pros/cons of using a browser.
5) openhmd - native for nothing, but supports [various things](https://github.com/OpenHMD/OpenHMD/wiki/Support-List) (if not very well). works on windows/mac/linux/android/bsd
6) libvr - exists. very basic oculus rift dk1/2/cv1 support on linux.

of these, the relevant ones are 1, 4, 2 if you want rift users to not hate you (or if you intend to ship on oculus store) and 3 if you are interested in gear vr/go. openvr and webvr net you the most hardware.

5 and 6 are pretty much irrelevant, although openhmd can net you basic psvr and rift cv1 support on platforms that wouldn't usually have that.

**Comment by bcampbell** [comment]
Just checked in a change to my branch which is a little better about the window being created after headset creation. And which doesn't coredump upon shutdown ;-)

Next up, redirected lovrHeadset functions, and runtime selection of implementation.
(the fake headset is probably also handy under emscripten, as a fallback for browsers which don't support webvr)

**Comment by bcampbell** [comment]
I've implemented runtime picking of fake/openvr/webvr implementations.
- the openvr one compiles fine, but I can't test it right now
- I've updated the webvr one, but haven't compiled it (no emscripten currently set up)

Could someone check out the openvr and webvr implementations for me please?
There shouldn't be any major problems, but bound to be a few minor things to fix up...

**Comment by bjornbytes** [comment]
OpenVR is working great for me.  I had to do add `#define _USE_MATH_DEFINES` before including math.h in openvr.c, and also had to rename every variable named near or far to clipNear/clipFar (so weird that I haven't run into this before!).

I'm going to try the WebVR build later today.

**Comment by bcampbell** [comment]
Great, thanks!
So far I've only been compiling it under Linux, which doesn't seem to have either of those two quirks. I did fix up a few near/far issues compiling for windows under gcc a while back, but I think it was only in the fake headset code - it was before the runtime switching.
Been a looooong time since I've seen near/far being used in the wild! Seems odd that they're still supported as compiler keywords, but I guess it lets old (16 bit segmented era very old) code compile...

**Comment by bjornbytes** [comment]
Only problem I saw with WebVR was that `glad.h` was getting included which caused some compile errors.  Normally it should be surrounded by `#ifndef EMSCRIPTEN`.  Looking at your branch it doesn't look like that was part of your change though, so I don't think there's anything to worry about!

Also, a heads up: while working on materials I changed `lovrControllerNewModelData` for OpenVR.  ModelData includes TextureData now so I was able to get rid of `lovrControllerNewTextureData` and put it in the ModelData.
